### ISSUE TITLE
Java class as inputType fails

### ISSUE THREAD

**Issue by jeggy** [aPureBase/KGraphQL#106]
Having a java class as an `inputType` will throw an `KotlinNullPointerException`.

Example setup:

```java
// Java
public class LatLng {
    public double lat;
    public double lng;

    public LatLng(double lat, double lng) {
        this.lat = lat;
        this.lng = lng;
    }
}
```
```kotlin
// Kotlin
fun main() {
    KGraphQL.schema {
        query("test") {
            resolver { radius: Double, location : LatLng ->
                "Hello $radius. ${location.lat} ${location.lng}"
            }
        }
    }.executeBlocking("""
        {
            test(radius: 2.3, location: { lat: 2.3, lng: 2.3 })
        }
    """).let(::println)
}
```

**Comment by Allali84** [comment]
Hello,
I tried to reproduce this bug but instead i have :

`kotlin.reflect.jvm.internal.KotlinReflectionInternalError: Cannot calculate JVM erasure for type: com.apurebase.kgraphql.bugs.LatLng`

I think this is normal because when i tried to force getting `KClass` from `kType` using the `javaType`, I got this new 
 exception :

`kotlin.reflect.jvm.internal.KotlinReflectionInternalError: Introspecting local functions, lambdas, anonymous functions and local variables is not yet fully supported in Kotlin reflection`

If I'm not mistaken, in our case it is "Introspecting anonymous functions"

Maybe we have to just throw an exception when a java class is used, it will be the simple solution

So what do you think ?

**Comment by jeggy** [comment]
I'm guessing you got that error because of having the java class within your kotlin directory?

But I'm thinking we should change this line https://github.com/aPureBase/KGraphQL/blob/master/kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/execution/ArgumentTransformer.kt#L29

So it should never throw a nullpointer exception but rather try to use reflection if no primary constructor is found and if that also is not possible, we should provide a good exception message explaining why it fails.

**Comment by Allali84** [comment]
Yes you are right, putting the java class in my kotlin directory was the cause of my error :P

So back to the problem of `NullPointerException` in this line :

https://github.com/aPureBase/KGraphQL/blob/master/kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/execution/ArgumentTransformer.kt#L29

We can get a list of constructors of the class and find the one that work for us (work both for java and kotlin), but the problem with java is that we can't get the parameter's name of constructors or any method (in java we have : arg0, arg1 ... unless you add **-parameters** option to the javac compiler)

I think throwing an adequate exception is the best solution at least for now
Something like : ExecutionException("Java class as inputType are not supported")

What do you think ?

**Comment by jeggy** [comment]
Yes, probably good to throw an exception like that for now, so we atleast don't get a `NullPointerException`.

But I also think it's doable to use reflection to instantiate a new object with reflection for an Java class. Could be done something like this:
If no Kotlin constructor is found we try to find the Java constructor and take the fields from that java constructor and loop through them and add our input values to an ArrayList _(which I think needs to be sorted correctly)_ and then call that constructor.

**Comment by Allali84** [comment]
Maybe I wasn't clear enough, what you said is possible but the problem is that in our case we absolutely need the parameter's name (e.g: `lat` and `lng`) because we need to map the values with the right parameter when we call the constructor (see line https://github.com/aPureBase/KGraphQL/blob/master/kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/execution/ArgumentTransformer.kt#L38)

and that not possible with java when we try to get the parameter we get `arg0` and `arg1` and not `lat` and `lng` (for more details about getting names of Method Parameters see here : https://docs.oracle.com/javase/tutorial/reflect/member/methodparameterreflection.html)

**Comment by jeggy** [comment]
This new improved warning about Java classes being unsupported has been added in version `0.15.2`.

I will leave this issue open as I still think we could solve it, maybe without using any constructor? But just instantiate a new object of this java class and then setting the values by mapping the fields after the class has been instantiated.
